src/index.js

```diff
@@ -19,664 +20,520 @@ const debug = (function () {
-      masqAppBaseUrl: options.masqAppBaseUrl ? options.masqAppBaseUrl : jsonConfig.masqAppBaseUrl,
-      swarmConfig: options.swarmConfig ? options.swarmConfig : jsonConfig.swarmConfig
-    }
+  get eventTarget () {
+    return this.service.machine.context.eventTarget
+  }
+
+  constructor (appName, appDescription, appImageURL, config) {
+    const eventTarget = document.createElement('MasqLib')
+    this.machine = Machine({
+      id: 'Masq',
+      initial: 'notInitialized',
+      context: {
+        appName,
+        appDescription,
+        appImageURL,
+        config,
+        eventTarget
+      },
+      strict: true,
+      on: {
+        SIGNALLING_SERVER_ERROR: 'loginFailed',
+        SW_CLOSED_DURING_LOGIN: 'loginFailed'
+      },
+      states: {
+        notInitialized: {
+          invoke: {
+            id: 'init',
+            src: _init,
+            onDone: {
+              target: 'notLogged',
+              actions: [ _sendReady ]
+            },
+            onError: {
+              target: 'initError'
+            }
+          }
+        },
+        initError: {
+        },
+        notLogged: {
+          on: {
+            LOGIN_START: 'loggingIn'
+          }
+        },
+        loggingIn: {
+          invoke: {
+            id: 'startingLogin',
+            src: _startLogin
+          },
+          on: {
+            DISCONNECTED: 'loginFailed',
+            AUTHORIZED: 'authorized',
+            NOT_AUTHORIZED: 'registerNeeded'
+          }
+        },
+        registerNeeded: {
+          invoke: {
+            src: _registerNeeded
+          },
+          on: {
+            MASQ_ACCESS_GRANTED: 'accessMaterialReceived',
+            MASQ_ACCESS_REFUSED: 'loginFailed'
+          }
+        },
+        accessMaterialReceived: {
+          invoke: {
+            src: _registering
+          },
+          on: {
+            WRITE_ACCESS_GRANTED: 'userAppDbCreated'
+          }
+        },
+        userAppDbCreated: {
+          invoke: {
+            src: _dbCreation
+          },
+          on: {
+            CONNECTION_ESTABLISHED: 'connectionEstablished'
+          }
+        },
+        authorized: {
+          invoke: {
+            src: _receivedAuthorized
+          },
+          on: {
+            CONNECTION_ESTABLISHED: 'connectionEstablished',
+            REGISTER_NEEDED: 'registerNeeded'
+          }
+        },
+        loginFailed: {
+          actions: [ _resetLogin ],
+          target: 'notLogged'
+        },
+        connectionEstablished: {
+          invoke: {
+            src: _readSessionInfoIntoState,
+            onDone: 'logged'
+          }
+        },
+        logged: {
+          on: {
+            SIGNOUT: {
+              target: 'notLogged',
+              actions: [_signout]
+            }
+          }
+        }
+      }
+    })
 
-    // init state
-    this.state = 'notLogged'
+    this.service = interpret(this.machine)
+
+    this.service.onTransition(state => {
```

Comment:
According to https://xstate.js.org/docs/guides/interpretation.html#transitions, we should add a condition IIUC `if (state.changed) { console.log(...) }`